cumulus/client/relay-chain-minimal-node/src/blockchain_rpc_client.rs

```diff
@@ -28,33 +28,43 @@ use polkadot_primitives::{
 	async_backing::{AsyncBackingParams, BackingState},
 	slashing, ApprovalVotingParams, CoreIndex, NodeFeatures,
 };
+use prometheus_endpoint::Registry;
 use sc_authority_discovery::{AuthorityDiscovery, Error as AuthorityDiscoveryError};
 use sc_client_api::AuxStore;
 use sp_api::{ApiError, RuntimeApiInfo};
 use sp_blockchain::Info;
 use sp_runtime::traits::{Block as BlockT, Header as HeaderT, NumberFor};
 
+use crate::metrics::BlockchainRpcMetrics;
+
 #[derive(Clone)]
 pub struct BlockChainRpcClient {
 	rpc_client: RelayChainRpcClient,
+	metrics: Option<BlockchainRpcMetrics>,
 }
 
 impl BlockChainRpcClient {
-	pub fn new(rpc_client: RelayChainRpcClient) -> Self {
-		Self { rpc_client }
+	pub fn new(rpc_client: RelayChainRpcClient, prometheus_registry: Option<Registry>) -> Self {
+		Self {
+			rpc_client,
+			metrics: prometheus_registry
+				.and_then(|registry| BlockchainRpcMetrics::register(&registry).ok()),
+		}
 	}
 
 	pub async fn chain_get_header(
 		&self,
 		hash: Option<Hash>,
 	) -> Result<Option<Header>, RelayChainError> {
+		let _timer = self.metrics.as_ref().map(|inner| inner.start_chain_get_header_timer());
```

Comment:
I don't understand the reasoning by exposing these timer APIs
Would be much easier to just to:

```rust
   let now = Instant::now();
   let res = self.rpc_client.chain_get_block_hash(number).await;
   self.metrics.as_ref().map(|m| m.chain_head(now));
   res
```

Reply:
or does this "timer" take the drop into account and emit the time on drop?

Reply:
> Would be much easier to just to:

```
   let now = Instant::now();
   let res = self.rpc_client.chain_get_block_hash(number).await;
   self.metrics.as_ref().map(|m| m.chain_head(now));
   res
```

I tried something along these line but the selling point of the timers API is that they simply return an f64 referring to seconds (with fractions of seconds included), which is observed behind the scenes by the underlying histogram (or at least that's my understanding, I still have to test it).

Using the Instant/Duration APIs looked more convoluted because of the calculations required to obtain that f64 expected by prometheus `observe` API (which requires getting the elapsed duration as seconds, but then `subsec_as_millis` maybe, and then convert their addition to f64).

Reply:
ok fair enough

Reply:
Using builtin timers looks super nice to me.